Source/Core/VideoCommon/TextureConverterShaderGen.cpp

```diff
@@ -147,156 +162,99 @@ ShaderCode GenerateShader(APIType api_type, const UidData* uid_data)
       break;
 
     case EFBCopyFormat::RA4:  // RA4
-      colmat[0] = colmat[4] = colmat[8] = colmat[15] = 1.0f;
-      color_mask[0] = color_mask[3] = 255.0f / 16.0f;
-      color_mask[4] = color_mask[7] = 1.0f / 15.0f;
-
-      if (!uid_data->efb_has_alpha)
-      {
-        color_mask[3] = 0.0f;
-        const_add[3] = 1.0f;
-      }
+      out.Write("  vec2 red_alpha = vec2(ivec2(texcol.ra * 255.0) & 0xF0) * (1.0 / 240.0);\n"
+                "  ocol0 = vec4(red_alpha.rrr, red_alpha.g);\n");
       break;
-    case EFBCopyFormat::RA8:  // RA8
-      colmat[0] = colmat[4] = colmat[8] = colmat[15] = 1.0f;
 
-      if (!uid_data->efb_has_alpha)
-      {
-        color_mask[3] = 0.0f;
-        const_add[3] = 1.0f;
-      }
+    case EFBCopyFormat::RA8:  // RA8
+      out.Write("  ocol0 = texcol.rrra;\n");
       break;
 
     case EFBCopyFormat::A8:  // A8
-      colmat[3] = colmat[7] = colmat[11] = colmat[15] = 1.0f;
-
-      if (!uid_data->efb_has_alpha)
-      {
-        color_mask[3] = 0.0f;
-        const_add[0] = 1.0f;
-        const_add[1] = 1.0f;
-        const_add[2] = 1.0f;
-        const_add[3] = 1.0f;
-      }
+      out.Write("  ocol0 = texcol.aaaa;\n");
       break;
 
     case EFBCopyFormat::G8:  // G8
-      colmat[1] = colmat[5] = colmat[9] = colmat[13] = 1.0f;
+      out.Write("  ocol0 = texcol.gggg;\n");
       break;
+
     case EFBCopyFormat::B8:  // B8
-      colmat[2] = colmat[6] = colmat[10] = colmat[14] = 1.0f;
+      out.Write("  ocol0 = texcol.bbbb;\n");
       break;
 
     case EFBCopyFormat::RG8:  // RG8
-      colmat[0] = colmat[4] = colmat[8] = colmat[13] = 1.0f;
+      out.Write("  ocol0 = texcol.rrrg;\n");
       break;
 
     case EFBCopyFormat::GB8:  // GB8
-      colmat[1] = colmat[5] = colmat[9] = colmat[14] = 1.0f;
+      out.Write("  ocol0 = texcol.gggb;\n");
       break;
 
     case EFBCopyFormat::RGB565:  // RGB565
-      colmat[0] = colmat[5] = colmat[10] = 1.0f;
-      color_mask[0] = color_mask[2] = 255.0f / 8.0f;
-      color_mask[4] = color_mask[6] = 1.0f / 31.0f;
-      color_mask[1] = 255.0f / 4.0f;
-      color_mask[5] = 1.0f / 63.0f;
-      const_add[3] = 1.0f;  // set alpha to 1
+      out.Write("  vec2 red_blue = vec2(ivec2(texcol.rb * 255.0) & 0xF8) * (1.0 / 248.0);\n"
+                "  float green = float(int(texcol.g * 255.0) & 0xFC) * (1.0 / 252.0);\n"
+                "  ocol0 = vec4(red_blue.r, green, red_blue.g, 1.0);\n");
       break;
 
     case EFBCopyFormat::RGB5A3:  // RGB5A3
-      colmat[0] = colmat[5] = colmat[10] = colmat[15] = 1.0f;
-      color_mask[0] = color_mask[1] = color_mask[2] = 255.0f / 8.0f;
-      color_mask[4] = color_mask[5] = color_mask[6] = 1.0f / 31.0f;
-      color_mask[3] = 255.0f / 32.0f;
-      color_mask[7] = 1.0f / 7.0f;
-
-      if (!uid_data->efb_has_alpha)
-      {
-        color_mask[3] = 0.0f;
-        const_add[3] = 1.0f;
-      }
+      out.Write("  vec3 color = vec3(ivec3(texcol.rgb * 255.0) & 0xF8) * (1.0 / 248.0);\n"
```

Comment:
May you add a TODO here about the correct RGB5A3? This format doesn't fit in 16 bit, but we can support the compressed rgb5a3 format here...